### DISCUSSION TITLE
Modbus support for STM32F4xx

### DISCUSSION THREAD

**Discussion by dresco** [terjeio/grblHAL#253] [General]
Just wondering, is there a plan for Modbus support in the STM32F4xx driver?

I ask because I'm fooling around with prototyping a Modbus based control panel, and had just started looking at plugin development for further testing. But the grblHAL board sitting on my desk is an F411 Nucleo, so didn't take me long to hit my first issue ;)

**Comment by terjeio** [comment]
The driver has a secondary UART channel that can be used for ModBus, and there is a lighweight plugin in the [spindle plugins](https://github.com/terjeio/grblHAL/tree/test/plugins/spindle) folder. This can be used for other purposes as well and enabling it should be easy as there is code for that present in the driver. Setting MODBUS_ENABLE to 1 in my_machine.h would do?

A full ModBus protocol implementation is not planned. However, I have ported the FreeModbus Library by Christian Walter to a couple of TI processors - this could be added as a plugin if a full implementation is required?

**Reply by dresco** [reply]
Thanks, yes I had been looking at the code for the spindle/modbus plugin, but then realised that plugin directory wasn't present for the STM32 driver. If you think it should work, I'll just copy it locally and see how I go..

**Reply by terjeio** [reply]
Yes, try it and check out the spindle plugin for how to interact with it.

**Reply by dresco** [reply]
Hmm, I've built the Huanyang spindle plugin with just a couple of minor changes to get it to compile (see below), but is getting stuck on the first modbus send.. 

The state never moves on from ModBus_TX, and I don't see any movement on the serial pins. I'm building for the Nucleo, so UART2 is actually USART1 on PA9 & 10

```
diff --git a/drivers/STM32F4xx/Inc/driver.h b/drivers/STM32F4xx/Inc/driver.h
index 9725f51..a88efef 100644
--- a/drivers/STM32F4xx/Inc/driver.h
+++ b/drivers/STM32F4xx/Inc/driver.h
@@ -200,6 +200,11 @@
   #endif
 #endif
 
+#if MODBUS_ENABLE
+  #include "modbus/modbus.h"
+  #define SERIAL2_MOD
+#endif
+
 // End configuration
 
 #if KEYPAD_ENABLE && !defined(KEYPAD_PORT)
diff --git a/drivers/STM32F4xx/Src/serial.c b/drivers/STM32F4xx/Src/serial.c
index 3bda5fe..a8ce304 100644
--- a/drivers/STM32F4xx/Src/serial.c
+++ b/drivers/STM32F4xx/Src/serial.c
@@ -292,9 +292,9 @@ void serial2Init (uint32_t baud_rate)
 
 #endif
 
-#if MODBUS_ENABLE
-    SERIAL2_MODULE->IE = EUSCI_A_IE_RXIE;
-#endif
+//#if MODBUS_ENABLE
+//    SERIAL2_MODULE->IE = EUSCI_A_IE_RXIE;
+//#endif
 }
 
 bool serial2SetBaudRate (uint32_t baud_rate)
```

**Reply by dresco** [reply]
> The state never moves on from ModBus_TX, and I don't see any movement on the serial pins. I'm building for the Nucleo, so UART2 is actually USART1 on PA9 & 10

Baby steps.. The Modbus plugin was loading garbage settings, and trying to set a nonsensical baud rate. With those reset to defaults, is now moving through the Modbus states correctly. Still not seeing any transitions on the actual TX pin though, even though is setting DR with the contents of the tx buffer..

**Reply by terjeio** [reply]
I have just uploaded some changes to the test branch, code was not finished... At least the TX interrupt fires so maybe it works now.

**Reply by dresco** [reply]
Thanks Terje, the strange thing is that I can see it the interrupt handler loading each byte from the modbus command into the DR data register, but the output on PA9 never toggles. (Have verified the hardware setup with a simple STMCube app that writes to USART1, which all works as expected)..

I'm left scratching my head now. If you get a chance, and you have a Nucleo F411 board, would you be able to check for the same behaviour at some point?

**Reply by terjeio** [reply]
Working for me - at least there is activity on the TX pin that looks good on the scope. Is your problem with the latest update? The previous version did not initialise the UART.

**Reply by dresco** [reply]
Thanks. Yep, I pulled the latest commits from the test branch & rebuilt, but is good to know it has to be something daft at my end. Will have another try tomorrow..

**Reply by dresco** [reply]
Aha, found my issue (eventually)! I was using the Generic Uno map, which assigns A9/A10 to stepper outputs..

**Comment by phil-barrett** [comment]
By the way, if you are looking at making changes for Modbus on the F411, the STM32F4xx UART has a half-duplex option that would fit well will RS-485 and save a pin.

**Reply by terjeio** [reply]
The modbus plugin supports half-duplex natively, by implementing/respecting the silence between messages. The driver needs to be modified if a direction pin is needed and a function for setting the direction output has to be bound to the `set_direction` function pointer in the modbus stream interface structure:

https://github.com/terjeio/grblHAL/blob/4683d7c86d71065f0abb0e7514a13610816d1880/plugins/spindle/modbus.h#L59-L71

**Reply by dresco** [reply]
I've ordered a couple of the cheapo modules with auto flow control from AliExpress for my PoC. In the post somewhere.. 

Eventually, I'd like to end up with a daughterboard for the serial header on the Teensy breakout. @phil-barrett, I'm ages away from that point still, but is the layout for the later boards available on your github? I was just after the mounting hole spacing that you'd added, but think I only found files for the earlier version..

**Reply by phil-barrett** [reply]
The newer layout is not currently available.  The serial daughter card mounting holes for the pro board are different.  I tried to keep them the same but wasn't able to.  Will be posting that and the I2C daughter card outlines/holes once I have completely finalized the Pro board (very close now).

For what it is worth, I have added a transmit direction pin to the serial daughter card header for the pro board.

**Reply by terjeio** [reply]
> For what it is worth, I have added a transmit direction pin to the serial daughter card header for the pro board.

Great, _driver.c_ should be updated to handle that when the pin is defined.

**Reply by phil-barrett** [reply]
Well, since the Teensy is maxed out on pins, I made it solder-bridge configurable for one of 4 possible pins: Aux2, Lim A, Lim B and KeyPad Strobe.  So, it will take a little config work in the pro map file, too.  

To be honest, the 4 pins were a guess. Maybe I should switch KPStrobe with Aux 1? I2C is getting more important, not less. Or one of the digital inputs?  Trying to go for maximum flexibility for different use cases.

I really like the STM32 half duplex feature - unfortunately, it doesn't look like iMXRTs have that.

**Reply by phomann** [reply]
Hi,
You can get around using a direction pin by having the UART TX pin of the micro drive the TX enable input on the RS-485 driver chip and then just tie the TX pin on the RS-485 chip to a logic 1. 

Cheers 

Peter

**Reply by terjeio** [reply]
> To be honest, the 4 pins were a guess. Maybe I should switch KPStrobe with Aux 1? I2C is getting more important, not less. Or one of the digital inputs? Trying to go for maximum flexibility for different use cases.

AUXINPUT0 (pin 36) can be used with little loss of optional functionality?

> You can get around using a direction pin by having the UART TX pin of the micro drive the TX enable input on the RS-485 driver chip and then just tie the TX pin on the RS-485 chip to a logic 1.

Would it not be better to get tranceivers with auto-direction circuitry - or is this the way it is usually implemented?
